### DISCUSSION TITLE
Java G-Code sender via Network (Telnet)

### DISCUSSION THREAD

**Discussion by timeO25** [winder/Universal-G-Code-Sender#2504] [Help]
Hello,
I want to make my own small G-Code sender with Java which sends the  G-Code via network (Telnet) to my microcontroller with GRBL on it. I read that the UGS is programmed in Java and also can send G-Code via network. The problem is I have no idea at the moment how to such a program looks like can someone help me with that?
Thank you very much.

**Comment by breiler** [comment]
A lot of work has been put into UGS to stream data to the controller efficiently. So we have different layers responsible for different things which is hard to just summarize.

To make the actual TCP connection to the controller you can have a look at this class:
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-core/src/com/willwinder/universalgcodesender/connection/TCPConnection.java

To handle a stream buffer you can have a look at this class:
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-core/src/com/willwinder/universalgcodesender/communicator/BufferedCommunicator.java

All events from the communicator to the UI must be sent asyncronysly which is done in this class:
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-core/src/com/willwinder/universalgcodesender/communicator/event/AsyncCommunicatorEventDispatcher.java

**Comment by timeO25** [comment]
I can now send the ? and then the response of the microcontroller is correct but when I try to send for example letters than grbl doesn't react. Maybe I encode the string wrong do you know how do I have to encode it?

**Reply by breiler** [reply]
No encoding should be needed. Did you end the sent line with a line break `\n`?

Read about the protocol here:
https://github.com/gnea/grbl/wiki/Grbl-v1.1-Interface

**Comment by timeO25** [comment]
Oh yeah I forgot that now it works thank you very much.
Do you coincidentally know how I have to send the 0x... commands to override the speed for example?

**Reply by breiler** [reply]
The byte is defined like this:
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-core/src/com/willwinder/universalgcodesender/GrblUtils.java#L69

Send it like this:
https://github.com/winder/Universal-G-Code-Sender/blob/master/ugs-core/src/com/willwinder/universalgcodesender/connection/TCPConnection.java#L137

**Comment by timeO25** [comment]
Thank you very much I got it working.
